### PR TITLE
ref(*): automatically plan and commit cluster changes (WIP)

### LINKED ISSUE #8: Automatically plan and commit the cluster
The current instructions for starting or scaling the cluster include two manual steps:
- Planning the cluster: `kubectl exec deis-riak-bootstrap riak-admin cluster plan`
- Committing the plan: `kubectl exec deis-riak-bootstrap riak-admin cluster commit`

In fact, these steps need to be done after any change in the Riak cluster topology, which includes a pod crashing and being restarted by Kubernetes. It's unacceptable for a cluster administrator to have to run those commands when they start or scale their Riak cluster, and especially so when Kubernetes automatically changes the topology. This component should take care of doing the plan and commit instead.

Ideas that should be tested:
1. Just run `riak-admin cluster plan && riak-admin cluster commit` after each `riak-admin cluster join` command (in nodes that are added to the cluster)
2. Build a simple API into the bootstrap pod that newly added pods can call with their node name. The API can then call the `cluster plan && cluster commit` calls in response. I'm not sure right now if this would be beneficial over option 1

### LINKED ISSUE #15: hook up 'go test' to CI
after #14, there are tests, so CI should run them

### PR DESCRIPTION
**Pull request by arschles**
Fixes #8 
Fixes #15 

This is a big patch, but the scope of it is to allow each individual Riak node to add itself completely to the cluster. The process for a node to add itself is as follows:
1. `riak-admin cluster join riak@$OTHER_RIAK_NODE_IP`
2. `riak-admin cluster plan`
3. `riak-admin cluster commit`

All 3 commands reconfigure the cluster, which:
1. Involves network communication to one or more other Riak nodes
2. Cause the cluster to un-converge (that's a word, right?)

As a result, this patch includes a rough, cluster-wide locking systme (to be enhanced in future PRs). The purpose of this system is to eliminate concurrent execution of these commands across nodes, and also to reduce the likelihood that Riak will reject the `plan` and `commit` commands because the cluster is not in a convergent state (it can still happen, though. See https://github.com/deis/riak/issues/18).

Details about the code herein:
- An HTTP server, called the "cluster server" that holds an in-memory lock and identifies it to callers by a UUID
- The original `boot` script has been split apart:
  - The original Riak start logic has been moved into the`start_riak` script
  - The logic to get the discovery service address and the `riak-admin cluster join` call has been ported to Go (it's in `./riak/join.go` now)
  - The aforementioned `./riak/join.go` code also grabs the cluster lock (from the cluster server mentioned above) and calls `riak-admin cluster plan && riak-admin cluster commit` when it has the lock. This is so that only one node does these cluster-wide operations at a time
  - Tests for all the new Go code have been added

### CODE REVIEW

File: clustersrv/client.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	"github.com/deis/riak/config"
+)
+
+func URLFromConfig(conf *config.Config) string {
+	return "http://" + conf.ClusterServerHost + ":" + strconv.Itoa(conf.ClusterServerPort)
```
Comment (technosophos) [inline clustersrv/client.go]: Lame nit: Using fmt.Sprintf() would save you from importing strconv.

Comment (arschles) [inline clustersrv/client.go]: Fair, change forthcoming

File: boot.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			}
+
+			if err := riak.Join(httpClient, clusterServerURL); err != nil {
+				cmdDoneCh <- err
+				return
+			}
+		}()
+	} else {
+		// bootstrap nodes should start (not join) a riak server and start the cluster server
+		log.Printf("Starting as a bootstrap node")
```
Comment (technosophos) [inline boot.go:46]: Duplicate of line 25. I think line 25 is supposed to say something else.

**Comment by bacongobbler** [comment]
smart use of the bootstrap pod to contain the lock server. :+1: on the implementation, too.

One question: if all new nodes boot are trying to run `cluster plan` and `cluster commit`, if another node as part of that batch runs plan & commit, is the other node's request essentially a no-op since the cluster's plan has already been applied? To dumb it down:
- I have a cluster of 3 nodes
- I add 3 more nodes
- all nodes join the cluster
- one acquires lock, the other two wait
- plan & commit

When one of the others acquire the lock next, will their plan & commit be a no-op or is there computation on Riak's end to essentially re-apply the same cluster plan?

File: boot.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			if err := riak.Start(); err != nil {
+				cmdDoneCh <- err
+				return
+			}
+		}()
+		log.Printf("Cluster server starting on port %d", conf.ClusterServerHTTPPort)
+		go clustersrv.Start(conf.ClusterServerHTTPPort, serverDoneCh)
+	}
+
+	if err := chans.JoinErrs(serverDoneCh, cmdDoneCh); err != nil {
```
Comment (technosophos) [inline boot.go:58]: If I read this right, if either the lock server or the riak server fails, the node will go down. What happens when the pod is brought back up? It will be a regular node and there will be no lock server?

File: boot.go (hunk 2 above)
Comment (arschles) [inline boot.go:46]: Yup, you're right

File: boot.go (hunk 3 above)
Comment (bacongobbler) [inline boot.go:58]: Pods have a restart policy AFAIK, so it will be brought back up as a bootstrap node with a lock server. I could be wrong about that though.

Comment (bacongobbler) [inline boot.go:58]: addressed with #17

**Comment by technosophos** [comment]
This is very, very clever. LGTM.

File: boot.go (hunk 3 above)
Comment (arschles) [inline boot.go:58]: @technosophos yea, you read it right. Further, if the bootstrap node goes down, a new node will be brought up by the replication controller, and it'll continuously crash because there'll no longer be a lock server available in the cluster.

https://github.com/deis/riak/issues/17 tracks that issue.

**Comment by arschles** [comment]
@bacongobbler good question, and sorry for the forthcoming wall of text.

I don't believe that subsequent `plan`/`commit` operations will be no-ops, because the `riak-admin cluster join riak@$DISCOVERY_IP` operation is also done inside the mutex. So, the ordering is this when adding 3 more nodes:
1. All pods come up
2. One acquires the lock
3. The pod from (2) runs `join`, `plan`, and `commit`
4. Repeat 2-3 for the other 2 nodes

However, there's wrinkle - after the `commit` and lock release, the Riak cluster state still hasn't necessarily reached convergence, so the `plan` operation (and hence the subsequent `commit`) can't complete. My understanding is that the Riak node in question can't get a clear picture of what vnodes it will be assigned (and where they will live on the ring). The probability that this happens increases along with the number of moving nodes (coming in or out of the ring). I captured the issue in https://github.com/deis/riak/issues/18, but I'm going to address it in a future PR.

File: boot.go (hunk 2 above)
Comment (arschles) [inline boot.go:46]: Fixed in https://github.com/arschles/deis-riak/commit/89befd16a7bd80741894d9289a90fb6c68c0d176
